resteasy-servlet-initializer/src/main/java/org/jboss/resteasy/plugins/servlet/ResteasyServletInitializer.java

```diff
@@ -95,51 +85,91 @@ protected void handleNoApplicationClass(Set<Class<?>> providers, Set<Class<?>> r
 
    }
 
+   private Set<ServletRegistration> getServletsForApplication(Class<?> applicationClass, ServletContext servletContext)
+   {
+      Set<ServletRegistration> set = new HashSet<>();
+      ServletRegistration reg = servletContext.getServletRegistration(applicationClass.getName());
+      if (reg != null && reg.getMappings().size() == 1)
+         set.add(reg);
+
+      for (ServletRegistration sr : servletContext.getServletRegistrations().values())
+      {
+         String appClassName = sr.getInitParameter(APPLICATION);
+         if (applicationClass.getName().equals(appClassName) && sr.getMappings().size() == 1)
+            set.add(sr);
+      }
+      return set;
+   }
 
    protected void register(Class<?> applicationClass, Set<Class<?>> providers, Set<Class<?>> resources, ServletContext servletContext)
    {
+      Set<ServletRegistration> servletsForApp = getServletsForApplication(applicationClass, servletContext);
+      // ignore @ApplicationPath if application is already mapped in web.xml
+      if (!servletsForApp.isEmpty())
+      {
+         for (ServletRegistration servletReg : servletsForApp)
+         {
+            String servletClassName = servletReg.getClassName();
+            if (servletClassName == null)
+               servletContext.addServlet(servletReg.getName(), HttpServlet30Dispatcher.class);
+            String prefix = servletReg.getMappings().iterator().next();
+            if (prefix.endsWith("*"))
+               prefix = prefix.substring(0, prefix.length() - 1);
+            if (prefix.length() > 1 && prefix.endsWith("/"))
+               prefix = prefix.substring(0, prefix.length() - 1);
+            registerResourcesAndProviders(servletReg, providers, resources, applicationClass, prefix);
+         }
+         return;
+      }
       ApplicationPath path = applicationClass.getAnnotation(ApplicationPath.class);

```

Comment:
The spec says
```
If the Application subclass is annotated with @ApplicationPath, implementations are REQUIRED to use
the value of this annotation appended with ”/*” to define a mapping for the added server.  Otherwise,
the application MUST be packaged with a web.xml that specifies a servlet mapping. 
```
Shouldn't this bit go first?

Reply:
Agreed.  My original reading of that led me to think that `@ApplicationPath` would be overridden if it was specified in the web.xml, but on further reflection I think your interpretation is correct.

In general, I don't like that approach as it means that a deployer has no way of changing the app path without re-coding/re-building the app. But in this case, it looks like that would be required.  I'll re-code this section.

Thanks for catching this!

Reply:
I added a new commit to ensure that the `@ApplicationPath` path is not overridden.

Reply:
At times like this I think of the song from Fiddler on the Roof: "Posing questions that would cross a rabbi's eyes". ;-)
  
The first part of register(),
```
      if (!servletsForApp.isEmpty())
      {
```
handles the case where a servlet is declared in web.xml for the Application. The spec says
```
If there is already a servlet that handles this application. That is, a servlet that has an initialization
parameter named  javax.ws.rs.Application whose value is the fully qualified name of the 
Application subclass, then no additional configuration steps are required by the JAX-RS implementation
```
That suggests that it's illegal to have a &lt;servlet&gt; element without a matching &lt;servlet-mapping&gt; element., although I haven't checked that. Also, it's saying, in effect, that @ApplicationPath has no role to play.

My interpretation is that the return statement you took out was correct. That is, if there IS a servlet, just use the servlet and matching &lt;servlet-mapping&gt; and return. Maybe it's worth checking for the existence of a &lt;servlet-mapping&gt;. It sounds like the JAX-RS spec is saying there must be a &lt;servlet-mapping&gt;, so it would be an error if there is none. But maybe that should get caught by the servlet container. I haven't checked that.

On the other hand, the spec says
```
If no servlet handles this  application, JAX-RS  implementations are REQUIRED  to dynamically
add a servlet whose fully qualified name must be that of the Application subclass.  If the Application
subclass is annotated with @ApplicationPath, implementations are REQUIRED to use the value of 
this annotation appended with ”/*” to define a mapping for the added server.  Otherwise, the application
MUST be packaged with a web.xml that specifies a servlet mapping. 
```
That suggests that it's legal to have a &lt;servlet-mapping&gt; element without a matching &lt;servlet&gt; element. So, if there's no declared servlet, and the Application doesn't have an @ApplicationPath, it sounds like we should then look for a &lt;servlet-mapping&gt; instead of giving up.

WDYT?

Reply:
And "if I were a rich man" then I could just retire and leave this issue to somebody else!  :-)

Ok, that spec text aligns with my previous understanding.  Use `@ApplicationPath` path if the path is not specified in the web.xml.  If it is specified in the web.xml, then ignore the `@ApplicationPath` path.  So, first off, I think I need to remove my last commit.

It looks like it is possible for a servlet to be defined in the web.xml without a mapping ( see https://stackoverflow.com/questions/12602805/can-we-invoke-a-servlet-without-servlet-mapping-in-web-xml-entry ).  The spec says that if an Application class is handled by a servlet, then no further configuration is necessary - unfortunately, it doesn't say one way or the other about whether the servlet must be mapped or not...

So, in the case where the servlet is mapped, I think we're fine.  None of my code is destructive if the user has specified things like the servlet class to handle it, etc. It does add RESTEASY init parameters (for scanned classes, etc.) - I don't think this should be a problem, but if we want to absolutely sure we could check that the servlet class name is a RESTEasy servlet before setting those parameters.

In the case where the application is handled by a servlet in the web.xml but does not have a mapping - I tend to think this is a bug - or best case using the jaws of life to remove a staple (using a JAX-RS application just to run something at app startup when a simple servlet would do).  The code as it is now (and as it will be after I remove the latest commit) would map the application to the `@ApplicationPath` path.  I think that this makes sense since we have no other path to use.  Another option would be to ignore the application and issue a warning to the user indicating that they should either delete the unmapped servlet from the web.xml or they should map it in the web.xml.  I'd be fine with either approach.

So:
1) Should we check the servlet class name to be sure it is a RESTEasy servlet before adding RESTEasy init parameters?
2) Should we ignore an application in the unmapped-servlet case and issue a warning?  Or default to the `@ApplicationPath` path?

and whichever way we go on 1 and 2, I will plan to revert the commit that removes the early return.

Thanks again!

Reply:
re: "I could just retire and leave this issue to somebody else"

Lolll. Don't tempt me!!

I will think about this some more tomorrow. :)

Reply:
This looks different every time I look at it. Holy Rashoman, Batman!!

I'm going to try to write out all of the possibilities, for my own understanding.

I. No Application

We MUST create a servlet, and there MUST be a &lt;servlet-mapping&gt; in the web.xml.  Technically, the Servlet spec would be ok without the &lt;servlet-mapping&gt;, but JAX-RS considers that an error.

II. Application exists

A. Matching servlet exists

The JAX-RS spec says, "no additional configuration steps are **required** [my emphasis]". However, the Servlet spec allows for the absence of a &lt;servlet-mapping&gt;, and JAX-RS doesn't FORBID using an @ApplicationPath path if it exists. In fact, it doesn't FORBID using "/\*" if @ApplicationPath doesn't exist. So that's different from my previous understanding. Actually, I think assuming we can use "/\*" as a default is probably a bad idea. It's probably better to consider it an error if there's no &lt;servlet-mapping&gt; and no @ApplicationPath.

B. Matching servlet doesn't exist

We MUST create a servlet. If @ApplicationPath exists, we MUST use it. Otherwise, the web.xml MUST have a matching &lt;servlet-mapping&gt;. [I agree that this feels backwards, btw.]

I think that's all consistent with your last comment, so that's good.

As for your questions:

1. If we find a servlet based on an Application name, then the servlet must be intended to be a JAX-RS servlet, right? So I think we're good, unless I'm missing something.

2. I like the idea of using the @ApplicationPath. The developer probably meant it to be used.

And finally, I think I've finally put my finger on something that's been bothering me semi-consciously. Suppose you have 

```
@ApplicationPath("/foo")
public class MyApp extends Application { ... }
```
If web.xml is
```
<web-app>
  <servlet>
    <servlet-name>MyApp</servlet-name>
  </servlet>
  <servlet-mapping>
    <servlet-name>MyApp</servlet-name>
    <url-pattern>/myresources/*</url-pattern>
  </servlet-mapping>
</web-app>
```
Then there **is** a matching servlet, and /myresources overrides /foo. But suppose web.xml is
```
<web-app>
  <servlet-mapping>
    <servlet-name>MyApp</servlet-name>
    <url-pattern>/myresources/*</url-pattern>
  </servlet-mapping>
</web-app>
```
Then, **I guess**, there **isn't** a matching servlet, and /foo overrides /myresources.

If I'm right, then
1. that's weird, and
2. probably &lt;servlet-mapping&gt; should override @ApplicationPath in both cases, for consistency.

If I'm right, maybe we should create an issue.

Reply:
@ronsigal everything looks right there.  

I don't think my changes handle case 1 (No Application subclass), but I don't think the previous code handled it either (looks like it was commented out...).  If we want to handle that case, I'd propose that we create an inner class like:
```
@ApplicationPath("/")
class DefaultApplication extends Application {}
```
and then add that class if there are no other Application subclasses in the web module.

As for the case where there is a servlet-mapping but not a servlet, that seems like a configuration error to me, but I'm checking with members of the web container team here at IBM to verify.  If a user is allowed to specify a servlet-mapping without a servlet, then I agree that if the servlet-mapping/servlet-name matches the Application subclass, then it makes sense to map that app subclass to the associated url-pattern.  That might be something we should clarify in the spec...

Please let me know whether you want me to handle these cases in this PR or a separate issue - either way works for me.  Thanks!

Reply:
Good morning. ;-)

1. I feel a little uncomfortable putting @ApplicationPath on DefaultApplication just because, if there **were** an Application, its @ApplicationPath would override the &lt;servlet-mapping&gt;.

2. I agree that having a &lt;servlet-mapping&gt; without a &lt;servlet&gt; seems strange, but otherwise I don't know how to make sense of 
```
If no servlet  handles  this  application,  JAX-RS  implementations  are  REQUIRED  to  dynamically
add a servlet ... .  If the Application subclass is annotated with @ApplicationPath, ... .  Otherwise, the application
MUST be packaged with a web.xml that specifies a servlet mapping.
```
3. Given the title of the JIRA, "Use servlet mapping from web.xml when not specified in @ApplicationPath", I think these new things can go here. The only thing I would break out is the code that allows a class to be both an Application and a resource, since that's more of a new feature.

Thanks!

Reply:
Sorry for the slow reply on this - I feel like I've been pulled in a hundred different directions since last week!

> 1. I feel a little uncomfortable putting @ApplicationPath on DefaultApplication just because, if there were an Application, its @ApplicationPath would override the <servlet-mapping>.

No problem.  The behavior of the no Application subclass case is the same in this change as it was before this change.  If we want to address that case further, I'd suggest we open a new JIRA for it.

> 2. I agree that having a <servlet-mapping> without a <servlet> seems strange, but otherwise I don't know how to make sense of

```
If no servlet  handles  this  application,  JAX-RS  implementations  are  REQUIRED  to  dynamically
add a servlet ... .  If the Application subclass is annotated with @ApplicationPath, ... .  Otherwise, the application
MUST be packaged with a web.xml that specifies a servlet mapping.
```

I think the spec clarifies it further using an example and table.  In the example of the web.xml it shows both a servlet and a servlet-mapping element.  The table says:
![Screen Shot 2020-09-24 at 8 59 01 AM](https://user-images.githubusercontent.com/21365299/94155140-3940ca00-fe44-11ea-9347-d68c7f9b0f97.png)

That says to me that the same web.xml content is required for an Application subclass without the `@ApplicationPath` as for the case where there is no Application subclass - and in that case too, the servlet and servlet-mapping are present in the web.xml.

My take on the text in question is that if there is no `@ApplicationPath` annotation, then the user MUST provide a web.xml that contains a servlet and servlet-mapping.  If there is an `@ApplicationPath` annotation, then the user does not need to provide a web.xml at all, but they can if they want to override the path specified in the application.

> 3. Given the title of the JIRA, "Use servlet mapping from web.xml when not specified in @ApplicationPath", I think these new things can go here. The only thing I would break out is the code that allows a class to be both an Application and a resource, since that's more of a new feature.

That works for me.  I'll split it out today and open a separate JIRA and PR for that feature.

Thanks again for all the review on this!